gio-subclass/tests/simple_action.rs

```diff
@@ -0,0 +1,391 @@
+            };
+            Box::new(imp)
+        }
+    }
+
+    impl ObjectImpl<Object> for SimpleAction {
+        fn set_property(&self, obj: &glib::Object, id: u32, value: &glib::Value) {
+            let prop = &PROPERTIES[id as usize];
+
+            match *prop {
+                Property::String("name", ..) => {
+                    self.name.replace(value.get().clone());
+                }
+                Property::Boolean("enabled", ..) => {
+                    self.enabled.replace(value.get().unwrap_or(false));
+                }
+                Property::Boxed("parameter-type", ..) => {
+                    self.parameter_type.replace(value.get().clone());
+                }
+                Property::Boxed("state-type", ..) => {
+                    self.state_type.replace(value.get().clone());
+                }
+                Property::Variant("state", ..) => {
+                    self.state.replace(value.get().clone());
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        fn get_property(&self, obj: &glib::Object, id: u32) -> Result<glib::Value, ()> {
+            let prop = &PROPERTIES[id as usize];
+
+            match *prop {
+                Property::String("name", ..) => Ok(self.name.borrow().clone().to_value()),
+                Property::Boolean("enabled", ..) => Ok(self.enabled.borrow().to_value()),
+                Property::Boxed("parameter-type", ..) => {
+                    Ok(self.parameter_type.borrow().clone().to_value())
+                }
+                Property::Boxed("state-type", ..) => {
+                    Ok(self.state_type.borrow().clone().to_value())
+                }
+                Property::Variant("state", ..) => Ok(self.state.borrow().clone().to_value()),
+                _ => unimplemented!(),
+            }
+        }
+    }
+
+    impl ActionImpl for SimpleAction {
+        fn activate(&self, action: &gio::Action, parameter: Option<&glib::Variant>) {
+
+            match *self.parameter_type.borrow(){
+                None => assert!(parameter.is_none()),
+                Some(ref t) => {
+                    assert!(parameter.is_some());
+                    assert!(parameter.unwrap().type_() == t);
+                }
+            };
+
+            if !*self.enabled.borrow() {
+                return;
+            }
+
+            /* If the user connected a signal handler then they are responsible
+             * for handling activation.
+             */
+
+            /* If not, do some reasonable defaults for stateful actions. */
+
+            match parameter{
+                Some(p) => { action.emit("activate", &[p]).unwrap(); },
+
+                None => {
+                    //FIXME: how can we propagate the optional nature of the arg?
+                    //       the value created here seems dangling for glib, and panics
+                    action.emit("activate", &[&glib::Variant::from(false)]).unwrap();
```

Comment:
You could pass `None` I guess

Reply:
Not really. The emit function is rather strict on its argument. It has to be something that implements to_value. And it has to be a ref.

Reply:
Wow. I can try that. I'm not even sure I understand what that means :)

Reply:
If you just write `&None` then the compiler tries to infer of what type `T` that `Option<T>` should be, but here it has no information at all what it could be. So you need to provide that type yourself. It's like with `let stuff = some_iter.collect::<Vec<_>>()` where you tell the compiler that you'd like to have a `Vec` and not something else that can be collected into